### PR TITLE
i2p: initial support

### PR DESCRIPTION
**Pull request by jtgrassie**
Adds i2p support on top of #4988.

### CODE REVIEW

File: ANONYMITY_NETWORKS.md [hunk 1]
```diff
@@ -165,15 +162,15 @@ simply a best effort attempt.
...
 
 #### Mitigation
 
-The best mitigiation is to use Kovri/I2P instead of Tor. However, Kovri/I2P
+The best mitigiation is to use I2P instead of Tor. However, I2P
 has a smaller set of users (less cover traffic) and academic reviews, so there
 is a tradeoff in potential isses. Also, anyone attempting this strategy really
 wants to uncover a user, it seems unlikely that this would be performed against
-every Tor/Kovri/I2P user.
+every Tor/I2P user.
```
Comment (moneromooo-monero) [inline ANONYMITY_NETWORKS.md]: You have a lot of diffs that just remove "Kovri" for no apparent reason.

File: src/net/i2p_address.cpp [hunk 2]
```diff
@@ -0,0 +1,200 @@
+// Copyright (c) 2018, The Monero Project
```
Comment (moneromooo-monero) [inline src/net/i2p_address.cpp]: 2019

File: src/net/i2p_address.cpp [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+        if (port_ != 0)
+        {
+            out.push_back(':');
+            namespace karma = boost::spirit::karma;
+            karma::generate(std::back_inserter(out), karma::ushort_, port());
+        }
+        return out;
+    }
+}
```
Comment (moneromooo-monero) [inline src/net/i2p_address.cpp]: This looks very much like the existing tor one. Can it be mostly merged ?

File: ANONYMITY_NETWORKS.md (hunk 1 above)
Comment (jtgrassie) [reply ANONYMITY_NETWORKS.md]: Kovri is an implementation, this is supposed to work with any i2p implementation. Just like Tor is not calling out particular implementations.

File: src/net/i2p_address.cpp (hunk 3 above)
Comment (jtgrassie) [reply src/net/i2p_address.cpp]: I considered that but their are other i2p addressing schemes too, so keeping separate should offer flexibility in adding those later on.

File: src/net/i2p_address.cpp (hunk 2 above)
Comment (jtgrassie) [reply src/net/i2p_address.cpp]: OK, well I started this in 2018, but changed anyway.

File: ANONYMITY_NETWORKS.md (hunk 1 above)
Comment (moneromooo-monero) [reply ANONYMITY_NETWORKS.md]: Good. I feared it might have been spite :)

**Comment by jtgrassie** [comment]
Rebased inline with https://github.com/monero-project/monero/pull/4988#issuecomment-458408689

**Comment by jtgrassie** [comment]
Rebased master now that #4988 is merged.

File: src/net/i2p_address.h [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        static const char* unknown_str() noexcept;
+
+        //! An object with `port() == 0` and `host_str() == unknown_str()`.
+        i2p_address() noexcept;
+
+        //! \return A default constructed `i2p_address` object.
+        static i2p_address unknown() noexcept { return i2p_address{}; }
+
+        /*!
+            Parse `address` in onion v2 or v3 format with (i.e. x.onion:80)
```
Comment (moneromooo-monero) [inline src/net/i2p_address.h]: That seems wrong.

File: src/net/i2p_address.h [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        ~i2p_address() = default;
+        i2p_address& operator=(const i2p_address& rhs) noexcept;
+
+        //! \return True if default constructed or via `unknown()`.
+        bool is_unknown() const noexcept;
+
+        bool equal(const i2p_address& rhs) const noexcept;
+        bool less(const i2p_address& rhs) const noexcept;
+
+        //! \return True if onion addresses are identical.
```
Comment (moneromooo-monero) [inline src/net/i2p_address.h]: That too.

File: src/net/i2p_address.h [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        bool is_same_host(const i2p_address& rhs) const noexcept;
+
+        //! \return `x.onion` or `x.onion:z` if `port() != 0`.
```
Comment (moneromooo-monero) [inline src/net/i2p_address.h]: And that.

File: src/net/i2p_address.h [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+        std::string str() const;
+
+        //! \return Null-terminated `x.onion` value or `unknown_str()`.
```
Comment (moneromooo-monero) [inline src/net/i2p_address.h]: Also that.

File: src/net/i2p_address.h (hunk 4 above)
Comment (jtgrassie) [reply src/net/i2p_address.h]: The comment is indeed wrong. Corrected this and the other comments.

**Pull request merged**